### PR TITLE
improve(std/asserts): allow assert function to specify type parameter

### PR DESCRIPTION
**Pull request by wongjiahau**
Such that user can have the option to ensure that they are comparing the right value.

This should be a better solution than #6376 to address #6375 as it can satisfy the requirements for both side of opposing view (i.e. looser types VS stronger types)

/cc @caspervonb @nayeemrmn

### CODE REVIEW

**Comment by wongjiahau** [comment]
@caspervonb Ok, I'm finding another way to allow this without using `any`

**Comment by caspervonb** [comment]
So I' not sure this would have any observable effect at all? As far as I know TypeScript will pick the lowest common denominator which would always be the `unknown`.

**Comment by wongjiahau** [comment]
I tried it, it does have effect. If you have time you can checkout this
branch to try it.

**Comment by wongjiahau** [comment]
@caspervonb For your reference:
image

Should I add a test to test the compile error? If so are there any defined convention in this project to test for compile errors?

**Comment by caspervonb** [comment]
Ah okay; i get it now explicit type parameters, missed that.

**Comment by wongjiahau** [comment]
Yeap, just like what the title of this PR says.

**Comment by wongjiahau** [comment]
Hi @ry can you help me review this?

**Comment by wongjiahau** [comment]
Hi @ry can you help me review this PR?

**Comment by ry** [comment]
Summoning @bartlomieju and @piscisaureus for their thoughts.

**Comment by wongjiahau** [comment]
Hi @piscisaureus ~ when you're free can you help me review this PR ?

**Comment by bartlomieju** [comment]
CC @kitsonk @nayeemrmn please take a look, I think having additional overloads with type parameters is fine, but want to get your opinions

**Comment by piscisaureus** [comment]
I think this is a good idea, but I have some questions:

* Is it actually necessary to *explicitly* state the type? I imagine that if I wrote `assertEquals(foo, 2)`, typescript would infer that `T = number` even if I didn't specify the type parameter.
* If the above is true, what's the purpose of keeping the overload with the `unknown` types?

**Comment by wongjiahau** [comment]
@piscisaureus Thanks for the review. I’ll do some changes to address your questions.

**Comment by nayeemrmn** [comment]
> * Is it actually necessary to _explicitly_ state the type? I imagine that if I wrote `assertEquals(foo, 2)`, typescript would infer that `T = number` even if I didn't specify the type parameter.

See all of the type errors that occur here: https://github.com/denoland/deno/runs/788255592.

@wongjiahau What changes? Are you just going to revert to #6376? I'm not for that one.

**Comment by wongjiahau** [comment]
@nayeemrmn No no I’m not going to revert the changes, instead I’ll make it break none of the test cases in this repo but still does not require user to explicitly specify the type parameters.

File: cli/tests/unit/timers_test.ts [hunk 1]
```diff
@@ -178,7 +178,7 @@ unitTest(async function timeoutCallbackThis(): Promise<void> {
   const { promise, resolve } = deferred();
   const obj = {
     foo(): void {
-      assertEquals(this, window);
+      assertEquals<unknown>(this, window);
```
Comment (wongjiahau) [inline cli/tests/unit/timers_test.ts:181]: This is the only case where casting to `unknown` is necessary under normal use cases.

File: std/testing/asserts_test.ts [hunk 2]
```diff
@@ -126,7 +126,7 @@ Deno.test("testingEqual", function (): void {
 Deno.test("testingNotEquals", function (): void {
   const a = { foo: "bar" };
   const b = { bar: "foo" };
-  assertNotEquals(a, b);
+  assertNotEquals<unknown>(a, b);
```
Comment (wongjiahau) [inline std/testing/asserts_test.ts:129]: The unknown casting in this file is only necessary because we are comparing values that obviously has different types.

File: std/testing/diff.ts [hunk 3]
```diff
@@ -6,11 +6,12 @@ interface FarthestPoint {
   id: number;
 }
 
-export enum DiffType {
-  removed = "removed",
-  common = "common",
-  added = "added",
-}
+export const DiffType = {
```
Comment (wongjiahau) [inline std/testing/diff.ts:9]: This change is necessary, because `string` is not assignable to enum types. Refer https://github.com/microsoft/TypeScript/issues/17690

File: std/testing/asserts.ts [hunk 4]
```diff
@@ -140,6 +140,27 @@ export function assert(expr: unknown, msg = ""): asserts expr {
...
+/**
+ * Check if a type is `any` type.
+ * Refer https://stackoverflow.com/a/55541672/6587634
+ */
+type IsAny<T> = T extends never ? true : never;
+type CheckNumberArray<T> = T extends NumberArray
+  ? NumberArray | undefined | null
+  : T | undefined | null;
+
+type ComparableType<T> = true extends IsAny<T> // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (nayeemrmn) [inline std/testing/asserts.ts:154]: This stuff is way too complex...

Comment (wongjiahau) [reply std/testing/asserts.ts:154]: Let me simplify it

File: std/testing/asserts.ts [hunk 5]
```diff
@@ -133,22 +133,35 @@ export function equal(c: unknown, d: unknown): boolean {
...
-/** Make an assertion, if not `true`, then throw. */
+/** Make an assertion, error will be thrown if `expr` does not have truthy value. */
 export function assert(expr: unknown, msg = ""): asserts expr {
   if (!expr) {
     throw new AssertionError(msg);
   }
 }
 
+type NumberArray = number[] | ArrayBuffer;
+type ComparableType<T> = T extends symbol
```
Comment (wongjiahau) [inline std/testing/asserts.ts:144]: @nayeemrmn Is this more comprehensible now?

File: std/testing/asserts.ts [hunk 6]
```diff
@@ -133,22 +133,35 @@ export function equal(c: unknown, d: unknown): boolean {
   })(c, d);
 }
 
-/** Make an assertion, if not `true`, then throw. */
+/** Make an assertion, error will be thrown if `expr` does not have truthy value. */
 export function assert(expr: unknown, msg = ""): asserts expr {
   if (!expr) {
     throw new AssertionError(msg);
   }
 }
```
Comment (wongjiahau) [inline std/testing/asserts.ts:142]: Is this more comprehensible now?

**Comment by wongjiahau** [comment]
> Is it actually necessary to explicitly state the type? I imagine that if I wrote assertEquals(foo, 2), typescript would infer that T = number even if I didn't specify the type parameter.

@piscisaureus I have applied some changes such that Typescript will infer the argument types for assert* functions. However, this might introduce a very tiny breaking changes (i.e. a few test cases requires type casting or specifying type parameter). Otherwise every other test cases can be type-checked properly without modifications. 

/cc @nayeemrmn @bartlomieju @kitsonk

File: std/node/events_test.ts [hunk 7]
```diff
@@ -382,7 +382,7 @@ Deno.test({
       ee.emit("event", 42, "foo");
     }, 0);
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const valueArr: any[] = await once(ee, "event");
+    const valueArr = (await once(ee, "event")) as [number, string];
```
Comment (wongjiahau) [inline std/node/events_test.ts:385]: If the value to be asserted has type of `any`, casting is also necessary

**Comment by piscisaureus** [comment]
@wongjiahau I really admire all the work you put into this, but I see now that my suggestion did not make complete sense, as assertEquals performs a loosely typed comparison. I think the resulting complexity is w/ ComparableType is undesirable.

I would like to keep the changes for`assertStrictEquals` but not for `assertEquals`.

**Comment by wongjiahau** [comment]
@piscisaureus May I also keep the changes for every other assert* functions? Because so far the only test cases that are broken by the stricter types are those that uses `assertEquals`

**Comment by piscisaureus** [comment]
I don't want to keep the `ComparableType` type.

For `assertStrictEquals` it shouldn't be necessary, if you keep just :

~`
export function assertStrictEquals<T>(
  actual: ComparableType<T>,
  expected: ComparableType<T>,
  msg?: string
): void;
`~
`
export function assertStrictEquals<T>(
  actual: T,
  expected: T,
  msg?: string
): void {
`

That should work because assertStrictEquals requires the compared values to have exactly the same type.

**Comment by wongjiahau** [comment]
@piscisaureus I have done what you requested, but just wondering why do you think `ComparableType` is undesirable? Because even with this type, almost every `assertEquals` usage wouldn't be affected negatively.

**Comment by piscisaureus** [comment]
@wongjiahau WDYT about https://github.com/denoland/deno/pull/6413/commits/5cd7aaab92016c37535e647524f4dc2cf4e3fdd3 ?

**Comment by piscisaureus** [comment]
https://github.com/denoland/deno/pull/6413/commits/ee2d89fecdbd9926077566957d8c1b9df9dd8432 is a very weird issue and I suspect there is an actually elegant solution, but I have no idea what's going on here...

@wongjiahau or @kitsonk maybe?

**Comment by wongjiahau** [comment]
@piscisaureus Regarding `ee2d89f`, this is one of the problem I faced when trying to create the `ComparableType` type. I found out that is actually working as intended, because by definition `unique symbol` is not assignable to `unique symbol`, as [discussed here](https://github.com/Microsoft/TypeScript/issues/23388).    
To solve this problem I used [this conditional type branch](https://github.com/denoland/deno/pull/6413/commits/058e066e11151a3a4149651ebc9e38e3e1673e80#diff-df7e00728a0711ee7b035dcfc46d1274R144) in `ComparableType`: 
```ts
T extends symbol ? symbol : ...
``` 

At the moment, the best solution I can think of is the solution above, which requires no explicit casting.

File: std/node/_util/_util_promisify.ts [hunk 8]
```diff
@@ -21,13 +21,25 @@
...
+declare const _CustomPromisifiedSymbol: unique symbol;
+declare const _CustomPromisifyArgsSymbol: unique symbol;
+
+declare let Symbol: SymbolConstructor;
+interface SymbolConstructor {
+  for(key: "nodejs.util.promisify.custom"): typeof _CustomPromisifiedSymbol;
+  for(
+    key: "nodejs.util.promisify.customArgs"
+  ): typeof _CustomPromisifyArgsSymbol;
+}
```
Comment (ry) [inline std/node/_util/_util_promisify.ts:33]: These changes seem unrelated?

Comment (piscisaureus) [reply std/node/_util/_util_promisify.ts:33]: No. https://github.com/denoland/deno/pull/6413/commits/ee2d89fecdbd9926077566957d8c1b9df9dd8432

**Comment by wongjiahau** [comment]
@piscisaureus Ok understood. Anyway thanks for merging!
